### ISSUE TITLE
RStudio Desktop 2023.12 and later overrides LD_LIBRARY_PATH.

### ISSUE THREAD

**Issue by bviviano** [rstudio/rstudio#15044]
### System details

    RStudio Edition : Desktop
    RStudio Version : 2023.12 and later
    OS Version      : Red Hat Enterprise Linux 8 or 9
    R Version       : 4.4.0 (but any version)

### Steps to reproduce the problem

Compile R with any non-standard compiler or auxiliary package that requires the use of `LD_LIBRARY_PATH` to locate those libraries.

### Describe the problem in detail

We are compiling our R software using the Intel OneAPI compilers, liking with the Intel Math Kernel Library to maximize performance.  Because of this, we need to set `LD_LIBRARY_PATH` to point to the location were the Intel libraries are located, as they are not installed in a standard location like `/usr/lib` or `/usr/lib64`.

When we launch RStudio with full DEBUG output, we see
```
2024-08-05T13:13:49.046Z DEBUG Ready to run
2024-08-05T13:13:49.169Z DEBUG instance lock: true, project: , file: false
2024-08-05T13:13:49.169Z DEBUG net-ipc: creating new message server; socket=/home/bviviano/.local/share/rstudio/rstudio.socket
2024-08-05T13:13:49.172Z DEBUG No rPath found, scanning for R
2024-08-05T13:13:49.173Z DEBUG Scanning system for R installations
2024-08-05T13:13:49.185Z DEBUG Using /usr/local/apps/R-4.4.0/intel-23.2/bin/R (found by /usr/bin/which/R)
2024-08-05T13:13:49.185Z DEBUG Done choosing R
2024-08-05T13:13:49.185Z DEBUG Preparing environment using R: /usr/local/apps/R-4.4.0/intel-23.2/bin/R
2024-08-05T13:13:49.186Z DEBUG Querying information about R executable at path: /usr/local/apps/R-4.4.0/intel-23.2/bin/R
2024-08-05T13:13:49.197Z DEBUG stdout: [no stdout produced]
2024-08-05T13:13:49.197Z DEBUG stderr: /usr/local/apps/R-4.4.0/intel-23.2/lib64/R/bin/exec/R: error while loading shared libraries: libiomp5.so: cannot open shared object file: No such file or directory

2024-08-05T13:13:49.197Z DEBUG status: 127 [failure]
2024-08-05T13:13:49.197Z DEBUG Error querying information about R: no output available
2024-08-05T13:13:49.199Z ERROR An unknown error occurred.
```
The specific issue being
```
2024-08-05T13:13:49.197Z DEBUG stderr: /usr/local/apps/R-4.4.0/intel-23.2/lib64/R/bin/exec/R: error while loading shared libraries: libiomp5.so: cannot open shared object file: No such file or directory
```
We traced this issue to `src/node/desktop/src/main/detect-r.ts`
```
  // remove R-related environment variables before invoking R
  const envCopy = Object.assign({}, process.env);
  delete envCopy['R_HOME'];
  delete envCopy['R_ARCH'];
  delete envCopy['R_DOC_DIR'];
  delete envCopy['R_INCLUDE_DIR'];
  delete envCopy['R_RUNTIME'];
  delete envCopy['R_SHARE_DIR'];
  delete envCopy[kLdLibraryPathVariable];
  delete envCopy['R_PLATFORM'];
```
Specifically the `delete envCopy[kLdLibraryPathVariable];` which seems to scrub out our `LD_LIBRARY_PATH` setting.  If we comment out that line and compile from source, RStudio Desktop works the same as it did with 2023.09.2 and earlier releases.

I traced the change to this specific commit https://github.com/rstudio/rstudio/commit/b113c3b5a21ecd4fba8281be4328510e5dde5d00

It seems like the intention of that commit was to change the order of `LD_LIBRARY_PATH`, but instead it is modifying it in such a way that it no longer works with our compiled versions or R.

I do not see an alternate method to pass in LD_LIBRARY_PATH into RStudio Desktop on Linux.  It seems like the use of `RSTUDIO_FALLBACK_LIBRARY_PATH` was perhaps an intended way of doing that, but looking at the code in `src/node/desktop/src/main/session-launcher.ts`, it appears to only apply to Apple Mac systems.

If I missed something in the docs or the code for doing the same for Linux or there is a known workaround, please let me know.

My issue seems to be similar to this open issue:

https://github.com/rstudio/rstudio/issues/14253

The above reported they were also trying to set `LD_LIBRARY_PATH` on their Linux systems to run a custom compiled R in a non-standard location.  I am opening a new ticket since that ticket not been updated in 6+ months and apears abandoned by the opener.

### Describe the behavior you expected

I expect to be able to set `LD_LIBRARY_PATH` before starting RStudio Desktop and have it pass unmodified to R so R can correctly locate all its libraries, like it did with RStudio Desktop 2023.09.2 and earlier.

- [x] I have read the guide for [submitting good bug reports](https://github.com/rstudio/rstudio/wiki/Writing-Good-Bug-Reports).
- [x] I have installed the latest version of RStudio, and confirmed that the issue still persists.
- [ ] If I am reporting an RStudio crash, I have included a [diagnostics report](https://support.posit.co/hc/en-us/articles/200321257-Running-a-Diagnostics-Report).
- [x] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

**Comment by ronblum** [comment]
@bviviano Thank you for raising the issue! I'll mark it for review.

**Comment by kevinushey** [comment]
Given that the failure here is happening when we try to invoke `R --vanilla`, I think the better solution here would be to make sure your R installation's `etc/ldpaths` sets the necessary library paths for your R session. This file is typically located at:

```
file.path(R.home("etc"), "ldpaths")
```

and is responsible for setting these library paths for the session. Can you check the contents of that file, and see if setting your library paths appropriately there resolves the issue?

**Comment by bviviano** [comment]
Whatever is set in `etc/ldpaths` would have no impact.  This is a failure of the `exec` system call to start R itself because it can't locate the Intel libraries because the `LD_LIBRARY_PATH` is changed before the attempted `exec` is called.

**Comment by kevinushey** [comment]
Are you sure? The program at `file.path(R.home("bin"), "R")` isn't actually an executable; it's a wrapper shell script that eventually calls `exec` to run R. That script is also responsible for sourcing the ldpaths script before R is started.

https://github.com/wch/r-source/blob/84a2587c4de6d8801d36f051e062a6de9f02bab9/src/scripts/R.sh.in#L250

**Comment by bviviano** [comment]
Sorry, you're right, I was thinking `etc/ldpath` was something getting called after R started.

Regardless, modifying `etc/ldpath` won't work for us.  We use lmod (https://lmod.readthedocs.io/en/latest/) on our HPC cluster to load and unload environment settings to control versions of R and versions of R packages based on specific user project needs.

Our user's need the flexibility to easily switch between say Intel 23.1, 23.2, 23.3, etc versions of the libraries and Intel Math Kernel Libraries because of subtle differences between them, without us needing to compile a version of R for each and every Intel compiler version installed.  So, we can't hard code specific Intel library paths into `etc/ldpath` without significant complexity (i.e. needing to compile, install and maintain multiple copies of the same R and extensions, for multiple compiler versions).

We really need a way to pass LD_LIBRARY_PATH through to R unadulterated in the way it did with RStudio 2023.9.2 and earlier.

**Comment by kevinushey** [comment]
Thanks for the extra context! That all sounds reasonable to me.

**Comment by gtritchie** [comment]
Verified that an existing LD_LIBRARY_PATH is no longer stripped when detecting R installations.
